### DISCUSSION TITLE
Language-aware (semantic) symbol search

### DISCUSSION THREAD

**Discussion by jglick** [community/community#8593] [Other Feature Feedback, Questions, & Ideas] [answered]
I see that https://cs.github.com/about/faq#languages supports “symbol extraction”. As far as I can tell, that means it does a lexical analysis of sources to find symbols (identifiers). But that is nowhere near adequate. Suppose I have an organization with hundreds of repositories in which one repository includes a `src/main/java/org/mine/api/Api.java`

```java
package org.mine.api;
public interface Api {
    void doSomething();
    void close(boolean force);
}
```

and another repository with `src/main/java/org/mine/impl/Impl.java`

```java
package org.mine.impl;
import org.mine.api.Api;
public class Impl implements Api, AutoCloseable {
    @Override public void doSomething() {…}
    @Override public void close() {
        close(false);
    }
    @Override public void close(boolean force) {…}
}
```

If I am exploring `Api.java`, searching for the symbol `close` is useless—I will get thousands of hits across the org, most of them totally unrelated, including one unrelated hit right in `Impl.java`. I want to know who is calling and/or implementing `org.mine.api.Api.close(boolean)` specifically. In a statically-typed language, it is possible to know the answer for sure (modulo weird build system tricks, reflection, etc.). [Sorcerer](https://github.com/kohsuke/sorcerer/blob/master/src/site/apt/index.apt) did this years ago for Java; you need to run an actual compiler front-end up to the symbol resolution phase (in a very lax mode suppressing errors), so that you can determine that a given identifier actually refers to a specific overload of a method in a specific type, or whatever the case may be.

**Comment by dcreager** [comment, marked as answer]
Hi @jglick, thanks for reaching out!  This is definitely on our radar.  We have just (as of 30 minutes ago) announced [_stack graphs_](https://github.blog/2021-12-09-precise-code-navigation-python-code-navigation-pull-requests/), which is our framework for extracting this kind of more detailed symbol information _without requiring a build_.  It's currently only enabled for Python, but we are hard at work adding stack graph support for other languages, as well.

**Reply by jglick** [reply]
Great news! I wish you best of luck interpreting Java’s symbol resolution rules without reusing an existing battle-hardened compiler (javac or JDT). The “hard” problem mentioned at the end of your talk (Java 5 syntax) actually seems fairly straightforward compared to the issues I have seen in practice. For the foreseeable future I expect search results should be treated as a best effort, suitable for casual exploration, with other tools used to answer definitive usage questions such as whether it is safe to delete a deprecated method.
